src/js/libgeo/GeoOps.js

```diff
@@ -616,15 +644,170 @@ geoOps.PointOnSegment.updatePosition = function(el) {
     var param = getStateComplexNumber();
     putStateComplexNumber(param); // copy parameter
     var seg = csgeo.csnames[el.args[0]];
-    var start = seg.startpos;
-    var end = seg.endpos;
-    var far = List.sub(end, start);
-    var homog = List.add(start, List.scalmult(param, far));
-    homog = List.normalizeMax(homog);
+
+    var homog = geoOps._helper.projectPointToSegment(seg, param);
     el.homog = General.withUsage(homog, "Point");
 };
 geoOps.PointOnSegment.stateSize = 2;
 
+geoOps._helper.PointOnArcCr = function(arc, P) {
+    var A = arc.startPoint,
+        B = arc.viaPoint,
+        C = arc.endPoint;
+    var cr = List.crossratio3harm(P, A, B, C, List.ii);
+
+    var m = cr.value[0];
+    var n = cr.value[1];
+
+    if (!CSNumber._helper.isAlmostZero(m)) {
+        n = CSNumber.div(n, m);
+        m = CSNumber.real(1);
+    } else {
+        m = CSNumber.div(m, n);
+        n = CSNumber.real(1);
+    }
+
+
+    var prod = m.value.real * n.value.real;
+
+    if (prod >= 0 && prod <= 1) {
+        var d1 = List.projectiveDistMinScal(P, A);
+        var d2 = List.projectiveDistMinScal(P, C);
```

Comment:
I wonder whether we want to take the closer point in terms of the `projectiveDistMinScal` metric, or whether we should use the already computed cross ratio for this. The same probably applies to point on segment as well.